js/app/widgets/dynamicLogo.js

```diff
@@ -0,0 +1,339 @@
+    },
+
+    _get_margin: function () {
+        let context = this._get_fresh_state_context();
+        let margin = context.get_margin(context.get_state());
+        context.restore();
+        return margin;
+    },
+
+    vfunc_get_request_mode: function () {
+        return Gtk.SizeRequestMode.CONSTANT_SIZE;
+    },
+
+    vfunc_get_preferred_width: function () {
+        let min_width = EosKnowledgePrivate.style_context_get_int(this.get_style_context(), 'min-width', this.get_state_flags());
+        min_width = min_width ? min_width : Width.MIN;
+        let nat_width = this._sizing === 'size-min' ? min_width : Width.MAX;
+
+        let margin = this._get_margin();
+        let extra = margin.left+ margin.right;
+
+        return [min_width + extra, nat_width + extra];
+    },
+
+    vfunc_get_preferred_height: function () {
+        let min_height = EosKnowledgePrivate.style_context_get_int(this.get_style_context(), 'min-height', this.get_state_flags());
+        min_height = min_height ? min_height : Height.MIN;
+        let nat_height = this._sizing === 'size-min' ? min_height : Height.MAX;
+
+        let margin = this._get_margin();
+        let extra = margin.top + margin.bottom;
+
+        return [min_height + extra, nat_height + extra];
+    },
+
+    vfunc_size_allocate: function (alloc) {
+        let margin = this._get_margin();
+        alloc.x += margin.left;
+        alloc.y += margin.top;
+        alloc.width -= margin.left + margin.right;
+        alloc.height -= margin.top + margin.bottom;
+        this.set_allocation(alloc);
+    },
+
+    vfunc_draw: function (cr) {
+        let width =  this.get_allocated_width();
+        let height = this.get_allocated_height();
+
+        if (this._previous_width !== width || this._previous_height !== height)
+            this._cache.invalidate();
+        this._previous_width = width;
+        this._previous_height = height;
+
+        cr.setSourceSurface(this._cache.get_surface(), 0, 0);
+        cr.paint();
+        cr.$dispose();
+    },
+
+    _draw_mode: function (cr) {
+        switch(this._mode) {
+            case 'image':
+                this._draw_image_mode(cr);
+                break;
+            case 'text':
+                this._draw_text_mode(cr);
+                break;
+            case 'full':
+                this._draw_full_mode(cr);
+        }
+    },
+
+    _draw_text_mode: function (cr) {
+        let context = this._get_fresh_state_context();
+        let total_width =  this.get_allocated_width();
+        let total_height = this.get_allocated_height();
+
+        let text_width = total_width > Width.MAX ? Width.MAX : total_width ;
+        let text_height = total_height > Height.MAX ? Height.MAX : total_height;
+
+        let font_desc = context.get_font(context.get_state());
+        let pango_layout = PangoCairo.create_layout(cr);
+        pango_layout.set_font_description(font_desc);
+        pango_layout.set_text(this.text, -1);
+
```

Comment:
`Gtk.Widget.create_pango_layout()` does not cache the `Pango.Layout`, though; you'll have to keep it around yourself in your class.